File: Classes/SBDocumentWindow.m
/*

SBDocumentWindow.m
 
Authoring by Atsushi Jike

Copyright 2010 Atsushi Jike. All rights reserved.

Redistribution and use in source and binary forms, with or without modification, 
are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer 
in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#import "SBDocumentWindow.h"
#import "SBBLKGUI.h"
#import "SBDefinitions.h"
#import "SBInnerView.h"
#import "SBSplitView.h"
#import "SBTabbar.h"
#import "SBUtil.h"

#import "Sunrise3-Bridging-Header.h"
#import "Sunrise3-Swift.h"

#define kSBFlipAnimationDuration 0.8
#define kSBFlipAnimationRectMargin 100
#define kSBBackWindowFrameWidth 800.0
#define kSBBackWindowFrameHeight 600.0

@implementation SBDocumentWindow

@dynamic innerRect;
@synthesize keyView;
@dynamic title;
@dynamic toolbar;
@dynamic contentView;
@synthesize innerView;
@synthesize coverWindow;
@synthesize tabbar;
@synthesize splitView;
@synthesize backWindow;
@synthesize tabbarVisivility;

- (instancetype)initWithFrame:(NSRect)frame delegate:(id)delegate tabbarVisivility:(BOOL)inTabbarVisivility
{
	NSUInteger styleMask = NSTitledWindowMask | NSClosableWindowMask |NSMiniaturizableWindowMask | NSResizableWindowMask;
	if (self = [super initWithContentRect:frame styleMask:styleMask backing:NSBackingStoreBuffered defer:YES])
	{
		[self constructInnerView];
        self.minSize = NSMakeSize(kSBDocumentWindowMinimumSizeWidth, kSBDocumentWindowMinimumSizeHeight);
        self.delegate = delegate;
        self.releasedWhenClosed = YES;
        self.showsToolbarButton = YES;
        self.oneShot = YES;
        self.acceptsMouseMovedEvents = YES;
        self.collectionBehavior = NSWindowCollectionBehaviorFullScreenPrimary | NSWindowCollectionBehaviorFullScreenAuxiliary;
        self.animationBehavior = NSWindowAnimationBehaviorNone;
		tabbarVisivility = inTabbarVisivility;
	}
	return self;
}

- (void)dealloc
{
    self.delegate = nil;
	[self destructCoverWindow];
}

- (BOOL)isCovering
{
	NSWindow *keyWindow = [NSApp keyWindow];
	return keyWindow ? keyWindow == coverWindow : NO;
}

- (BOOL)canBecomeKeyWindow
{
	BOOL r = self.covering ? NO : super.canBecomeKeyWindow;
	return r;
}

- (void)becomeKeyWindow
{
	[super becomeKeyWindow];
	if (coverWindow)
	{
		[coverWindow makeKeyWindow];
	}
}

#pragma mark Getter

- (NSString *)title
{
	return super.title;
}

- (SBToolbar *)toolbar
{
	return (SBToolbar *)super.toolbar;
}

- (NSView *)contentView
{
	return super.contentView;
}

#pragma mark Rects

- (NSRect)innerRect
{
	return self.contentView.bounds;
}

- (CGFloat)tabbarHeight
{
	return kSBTabbarHeight;
}

- (NSRect)tabbarRect
{
	NSRect r = NSZeroRect;
	NSRect innerRect = self.innerRect;
	r.size.width = innerRect.size.width;
	r.size.height = self.tabbarHeight;
	r.origin.y = tabbarVisivility ? innerRect.size.height - r.size.height : innerRect.size.height;
	return r;
}

- (NSRect)splitViewRect
{
	NSRect r = NSZeroRect;
	NSRect innerRect = self.innerRect;
	r.size.width = innerRect.size.width;
	r.size.height = tabbarVisivility ? innerRect.size.height - self.tabbarHeight : innerRect.size.height;
	return r;
}

- (CGFloat)sheetPosition
{
	CGFloat position = self.splitViewRect.size.height;
	return position;
}

#pragma mark Responding

- (BOOL)performKeyEquivalent:(NSEvent *)theEvent
{
	BOOL r = NO;
	if (self.delegate)
	{
		if ([self.delegate respondsToSelector:@selector(window:shouldHandleKeyEvent:)])
		{
			r = [(id<SBDocumentWindowDelegate>)self.delegate window:self shouldHandleKeyEvent:theEvent];
		}
	}
	if (!r)
	{
		r = [super performKeyEquivalent:theEvent];
	}
	return r;
}

#pragma mark Actions

- (void)performClose:(id)sender
{
	BOOL shouldClose = YES;
	id delegate = self.delegate;
	if (delegate)
	{
		if ([delegate respondsToSelector:@selector(window:shouldClose:)])
		{
			shouldClose = [delegate window:self shouldClose:sender];
		}
	}
	if (shouldClose)
	{
		[super performClose:sender];
	}
}

#pragma mark Construction

- (void)constructInnerView
{
	innerView = [[SBInnerView alloc] initWithFrame:self.innerRect];
    innerView.autoresizingMask = NSViewWidthSizable | NSViewHeightSizable;
	[self.contentView addSubview:innerView];
}

#pragma mark Menu validation

- (BOOL)validateMenuItem:(NSMenuItem *)menuItem
{
	BOOL r = YES;
	SEL selector = menuItem.action;
	if (selector == @selector(toggleToolbarShown:))
	{
        menuItem.title = self.toolbar.visible ? NSLocalizedString(@"Hide Toolbar", nil) : NSLocalizedString(@"Show Toolbar", nil);
		r = !self.coverWindow;
	}
	else {
		r = [super validateMenuItem:menuItem];
	}
	return r;
}

#pragma mark Setter

- (void)setTitle:(NSString *)title
{
	if (!title)
	{
		title = [NSString string];
	}
    super.title = title;
}

- (void)setToolbar:(SBToolbar *)toolbar
{
	if (self.toolbar != toolbar)
	{
        super.toolbar = toolbar;
	}
}

- (void)setContentView:(NSView *)contentView
{
    super.contentView = contentView;
}

- (void)setTabbar:(SBTabbar *)inTabbar
{
	if (tabbar != inTabbar)
	{
		tabbar = inTabbar;
		NSRect r = self.tabbarRect;
		if (!NSEqualRects(tabbar.frame, r))
		{
			tabbar.frame = r;
		}
        tabbar.autoresizingMask = NSViewWidthSizable | NSViewMinYMargin;
		[self.innerView addSubview:tabbar];
	}
}

- (void)setSplitView:(SBSplitView *)inSplitView
{
	if (splitView != inSplitView)
	{
		splitView = inSplitView;
		NSRect r = self.splitViewRect;
		if (!NSEqualRects(splitView.frame, r))
		{
            splitView.frame = r;
		}
        splitView.autoresizingMask = NSViewWidthSizable | NSViewHeightSizable;
		[self.innerView addSubview:(NSView *)splitView];
	}
}

- (void)setKeyView:(BOOL)isKeyView
{
	if (keyView != isKeyView)
	{
		keyView = isKeyView;
	}
}

#pragma mark Actions

- (void)zoom:(id)sender
{
	if (!coverWindow)
	{
		[super zoom:sender];
	}
}

- (void)destructCoverWindow
{
	if (coverWindow)
	{
		[self removeChildWindow:coverWindow];
		[coverWindow close];
		coverWindow = nil;
	}
    self.showsToolbarButton = YES;
}

- (void)showCoverWindow:(SBView *)view
{
	NSRect r = view.frame;
	NSSize size = self.innerRect.size;
	r.origin.x = (size.width - r.size.width) / 2;
	r.origin.y = (size.height - r.size.height) / 2;
	view.frame = r;
	[self constructCoverWindowWithView:view];
}

- (void)constructCoverWindowWithView:(NSView *)view
{
	SBBLKGUIScrollView *scrollView = nil;
	NSRect vr = view.frame;
	NSRect br = self.splitView.bounds;
	NSRect r = NSZeroRect;
	BOOL hasHorizontalScroller = vr.size.width > br.size.width;
	BOOL hasVerticalScroller = vr.size.height > br.size.height;
	r.origin.x = hasHorizontalScroller ? br.origin.x : vr.origin.x;
	r.size.width = hasHorizontalScroller ? br.size.width : vr.size.width;
	r.origin.y = hasVerticalScroller ? br.origin.y : vr.origin.y;
	r.size.height = hasVerticalScroller ? br.size.height : vr.size.height;
	[self destructCoverWindow];
	coverWindow = [[SBCoverWindow alloc] initWithParentWindow:self size:br.size];
	scrollView = [[SBBLKGUIScrollView alloc] initWithFrame:NSIntegralRect(r)];
    scrollView.autoresizingMask = NSViewWidthSizable | NSViewHeightSizable;
    scrollView.hasHorizontalScroller = hasHorizontalScroller;
    scrollView.hasVerticalScroller = hasVerticalScroller;
    scrollView.drawsBackground = NO;
	[coverWindow.contentView addSubview:scrollView];
    coverWindow.releasedWhenClosed = NO;
    scrollView.documentView = view;
    self.showsToolbarButton = NO;
	
#if 1
	[self addChildWindow:coverWindow ordered:NSWindowAbove];
	[coverWindow makeKeyWindow];
#else
	NSViewAnimation *animation = nil;
	NSDictionary *animationInfo = nil;
    coverWindow.contentView.hidden = YES;
	[NSApp beginSheet:coverWindow modalForWindow:self modalDelegate:self didEndSelector:@selector(coverWindowDidEnd:returnCode:contextInfo:) contextInfo:nil];
    
    animationInfo = @{NSViewAnimationTargetKey: coverWindow.contentView,
                      NSViewAnimationEffectKey: NSViewAnimationFadeInEffect};
	animation = [[NSViewAnimation alloc] initWithViewAnimations:@[animationInfo]];
    animation.duration = 0.35;
    animation.animationBlockingMode = NSAnimationNonblockingThreaded;
    animation.animationCurve = NSAnimationEaseIn;
    animation.delegate = self;
	[animation startAnimation];
    coverWindow.contentView.hidden = NO;
#endif
}

#if 1
- (void)hideCoverWindow
{
	[self removeChildWindow:coverWindow];
	[coverWindow orderOut:nil];
	[self destructCoverWindow];
	[self makeKeyWindow];
}
#else
- (void)animationDidStop:(NSAnimation *)animation
{
	if (coverWindow.contentView)
	{
        coverWindow.contentView.hidden = NO;
	}
}

- (void)coverWindowDidEnd:(NSWindow *)window returnCode:(NSInteger)returnCode contextInfo:(void *)contextInfo
{
	[self destructCoverWindow];
}

- (void)hideCoverWindow
{
	[NSApp endSheet:coverWindow];
}
#endif

- (void)hideToolbar
{
	if (self.toolbar.visible)
		[self toggleToolbarShown:self];
}

- (void)showToolbar
{
    if (self.toolbar.visible)
        [self toggleToolbarShown:self];
}

- (void)hideTabbar
{
	if (tabbarVisivility)
	{
		tabbarVisivility = NO;
		NSRect r = NSZeroRect;
		r = self.tabbarRect;
		if (!NSEqualRects(tabbar.frame, r))
		{
			tabbar.frame = r;
		}
		r = self.splitViewRect;
		if (!NSEqualRects(splitView.frame, r))
		{
			splitView.frame = r;
		}
	}
}

- (void)showTabbar
{
	if (!tabbarVisivility)
	{
		tabbarVisivility = YES;
		NSRect r = NSZeroRect;
		r = self.tabbarRect;
		if (!NSEqualRects(tabbar.frame, r))
		{
			tabbar.frame = r;
		}
		r = self.splitViewRect;
		if (!NSEqualRects(splitView.frame, r))
		{
			splitView.frame = r;
		}
	}
}

- (void)flip
{
	SBBLKGUIButton *doneButton = nil;
	NSRect doneRect = NSZeroRect;
	doneRect.size.width = 105.0;
	doneRect.size.height = 24.0;
	doneButton = [[SBBLKGUIButton alloc] initWithFrame:doneRect];
    doneButton.title = NSLocalizedString(@"Done", nil);
    doneButton.target = self;
    doneButton.action = @selector(doneFlip);
    doneButton.enabled = YES;
	doneButton.keyEquivalent = @"\r";
	[self flip:(SBView *)doneButton];
}

- (void)flip:(SBView *)view
{
	NSRect r = self.frame;
	NSRect br = r;
	br.size.width = kSBBackWindowFrameWidth;
	br.size.height = kSBBackWindowFrameHeight;
	br.origin.x = self.frame.origin.x + (self.frame.size.width - br.size.width) / 2;
	br.origin.y = self.frame.origin.y + (self.frame.size.height - br.size.height) / 2;
	br.size.height -= 23.0;
	backWindow = [[NSWindow alloc] initWithContentRect:br styleMask:(NSTitledWindowMask | NSClosableWindowMask) backing:NSBackingStoreBuffered defer:YES];
	backWindow.backgroundColor = [NSColor colorWithCalibratedRed:SBWindowBackColors[0] green:SBWindowBackColors[1] blue:SBWindowBackColors[2] alpha:SBWindowBackColors[3]];
    backWindow.releasedWhenClosed = NO;
	view.frame = NSMakeRect((br.size.width - view.frame.size.width) / 2, (br.size.height - view.frame.size.height) / 2, view.frame.size.width, view.frame.size.height);
	[backWindow.contentView addSubview:view];
	[backWindow makeKeyAndOrderFront:nil];
    self.alphaValue = 0;
}

- (void)doneFlip
{
	if (backWindow)
	{
		[backWindow close];
		backWindow = nil;
	}
    self.alphaValue = 1;
}

@end

File: Classes/SBDocumentWindow.swift
/*
SBDocumentWindow.swift

Copyright (c) 2014, Alice Atlas
Copyright (c) 2010, Atsushi Jike
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

@objc
protocol SBDocumentWindowDelegate: NSWindowDelegate {
    optional func window(window: SBDocumentWindow, shouldClose: AnyObject?) -> Bool
    optional func window(window: SBDocumentWindow, shouldHandleKeyEvent: NSEvent) -> Bool
    optional func windowDidFinishFlipping(window: SBDocumentWindow)
}

class SBDocumentWindow: NSWindow {
    let kSBFlipAnimationDuration: CGFloat = 0.8
    let kSBFlipAnimationRectMargin: CGFloat = 100
    let kSBBackWindowFrameWidth: CGFloat = 800.0
    let kSBBackWindowFrameHeight: CGFloat = 600.0

	var backWindow: NSWindow?
	var keyView = false
	lazy var innerView: SBInnerView = {
        let innerView = SBInnerView(frame: self.innerRect)
        innerView.autoresizingMask = .ViewWidthSizable | .ViewHeightSizable
        return innerView
    }()
	var coverWindow: SBCoverWindow?
    var tabbar: SBTabbar? {
        didSet {
            oldValue?.removeFromSuperview()
            if tabbar != nil {
                tabbar!.frame = tabbarRect
                tabbar!.autoresizingMask = .ViewWidthSizable | .ViewMinYMargin
                innerView.addSubview(tabbar!)
            }
        }
    }
	var tabbarVisibility: Bool = false
    var sbToolbar: SBToolbar? {
        get { return super.toolbar as? SBToolbar }
        set(sbToolbar) { super.toolbar = sbToolbar }
    }
    var sbDelegate: SBDocumentWindowDelegate? {
        get { return delegate as? SBDocumentWindowDelegate }
        set(sbDelegate) { delegate = sbDelegate }
    }
    var splitView: SBSplitView? {
        didSet {
            oldValue?.removeFromSuperview()
            if splitView != nil {
                splitView!.frame = splitViewRect
                splitView!.autoresizingMask = .ViewWidthSizable | .ViewMinYMargin
                innerView.addSubview(splitView!)
            }
        }
    }
    override var title: String! {
        didSet {
            let title: String? = self.title
            super.title = (title ?? "") as String
        }
    }

    init(frame: NSRect, delegate: SBDocumentWindowDelegate?, tabbarVisibility inTabbarVisibility: Bool) {
        let styleMask = NSTitledWindowMask | NSClosableWindowMask | NSMiniaturizableWindowMask | NSResizableWindowMask
        super.init(contentRect: frame, styleMask: styleMask, backing: .Buffered, defer: true)
        
        contentView.addSubview(innerView)
        minSize = NSMakeSize(kSBDocumentWindowMinimumSizeWidth, kSBDocumentWindowMinimumSizeHeight)
        sbDelegate = delegate
        releasedWhenClosed = true
        showsToolbarButton = true
        oneShot = true
        acceptsMouseMovedEvents = true
        collectionBehavior = .FullScreenPrimary | .FullScreenAuxiliary
        animationBehavior = .None
        tabbarVisibility = inTabbarVisibility
    }
    
    required init(coder: NSCoder) {
        fatalError("NSCoding not supported")
    }
    
    deinit {
        sbDelegate = nil
        destructCoverWindow()
    }
    
    var covering: Bool {
        if let keyWindow = NSApplication.sharedApplication().keyWindow {
            return keyWindow === coverWindow
        }
        return false
    }
    
    override var canBecomeKeyWindow: Bool {
        return covering ? false : super.canBecomeKeyWindow
    }
    
    override func becomeKeyWindow() {
        super.becomeKeyWindow()
        coverWindow?.makeKeyWindow()
    }
    
    // MARK: Rects
    
    var innerRect: NSRect { return contentView.bounds }
    let tabbarHeight = kSBTabbarHeight
    
    var tabbarRect: NSRect {
        var r = NSZeroRect
        r.size.width = innerRect.size.width
        r.size.height = tabbarHeight
        r.origin.y = tabbarVisibility ? innerRect.size.height - r.size.height : innerRect.size.height
        return r
    }
    
    var splitViewRect: NSRect {
        var r = NSZeroRect
        r.size.width = innerRect.size.width
        r.size.height = tabbarVisibility ? innerRect.size.height - tabbarHeight : innerRect.size.height
        return r
    }
    
    var sheetPosition: CGFloat { return splitViewRect.size.height }
    
    // MARK: Responding
    
    override func performKeyEquivalent(event: NSEvent) -> Bool {
        if sbDelegate?.window?(self, shouldHandleKeyEvent: event) ?? false {
            return true
        }
        return super.performKeyEquivalent(event)
    }
    
    // MARK: Actions
    
    override func performClose(sender: AnyObject?) {
        if sbDelegate?.window?(self, shouldClose: sender) ?? true {
            super.performClose(sender)
        }
    }
    
    // MARK: Menu validation
    
    override func validateMenuItem(menuItem: NSMenuItem) -> Bool {
        if menuItem.action == "toggleToolbarShown:" {
            menuItem.title = toolbar.visible ? NSLocalizedString("Hide Toolbar", comment: "") : NSLocalizedString("Show Toolbar", comment: "")
            return coverWindow == nil
        }
        return super.validateMenuItem(menuItem)
    }
    
    // MARK: Actions
    
    override func zoom(sender: AnyObject?) {
        if coverWindow == nil {
            super.zoom(sender)
        }
    }
    
    func destructCoverWindow() {
        if coverWindow != nil {
            removeChildWindow(coverWindow!)
            coverWindow!.close()
            coverWindow = nil
        }
        showsToolbarButton = true
    }
    
    func showCoverWindow(view: SBView) {
        var r = view.frame
        let size = innerRect.size
        r.origin.x = (size.width - r.size.width) / 2
        r.origin.y = (size.height - r.size.height) / 2
        view.frame = r
        constructCoverWindowWithView(view)
    }
    
    func constructCoverWindowWithView(view: NSView) {
        let vr = view.frame
        let br = splitView!.bounds
        var r = NSZeroRect
        let hasHorizontalScroller = vr.size.width > br.size.width
        let hasVerticalScroller = vr.size.height > br.size.height
        r.origin.x = hasHorizontalScroller ? br.origin.x : vr.origin.x
        r.size.width = hasHorizontalScroller ? br.size.width : vr.size.width
        r.origin.y = hasVerticalScroller ? br.origin.y : vr.origin.y
        r.size.height = hasVerticalScroller ? br.size.height : vr.size.height
        destructCoverWindow()
        coverWindow = SBCoverWindow(parentWindow: self, size: br.size)
        let scrollView = SBBLKGUIScrollView(frame: NSIntegralRect(r))
        scrollView.autoresizingMask = .ViewWidthSizable | .ViewHeightSizable
        scrollView.hasHorizontalScroller = hasHorizontalScroller
        scrollView.hasVerticalScroller = hasVerticalScroller
        scrollView.drawsBackground = false
        coverWindow!.contentView.addSubview(scrollView)
        coverWindow!.releasedWhenClosed = false
        scrollView.documentView = view
        self.showsToolbarButton = false
        
        #if true
            addChildWindow(coverWindow, ordered: .Above)
            coverWindow!.makeKeyWindow()
        #else
            coverWindow.contentView.hidden = true
            NSApp.beginSheet(coverWindow, modalForWindow: self, modalDelegate: self, didEndSelector: "coverWindowDidEnd:returnCode:contextInfo:", contextInfo: nil)
            
            animation = NSViewAnimation(viewAnimations: [[
                NSViewAnimationTargetKey: coverWindow.contentView,
                NSViewAnimationEffectKey: NSViewAnimationFadeInEffect]])
            animation.duration = 0.35
            animation.animationBlockingMode = .NonblockingThreaded
            animation.animationCurve = .EaseIn
            animation.delegate = self
            animation.startAnimation()
            coverWindow.contentView.hidden = false
        #endif
    }

    #if true
        func hideCoverWindow() {
            removeChildWindow(coverWindow)
            coverWindow!.orderOut(nil)
            destructCoverWindow()
            makeKeyWindow()
        }
    #else
        func animationDidStop(animation: NSAnimation) {
            if coverWindow?.contentView != nil {
                coverWindow!.contentView!.hidden = false
            }
        }
        
        func coverWindowDidEnd(window: NSWindow, returnCode: Int, contextInfo: AnyObject?) {
            destructCoverWindow()
        }
        
        func hideCoverWindow() {
            NSApp.endSheet(coverWindow)
        }
    #endif
    
    func hideToolbar() {
        if toolbar.visible {
            toggleToolbarShown(self)
        }
    }
    
    func showToolbar() {
        if !toolbar.visible {
            toggleToolbarShown(self)
        }
    }
    
    func hideTabbar() {
        if tabbarVisibility {
            tabbarVisibility = false
            tabbar!.frame = tabbarRect
            splitView!.frame = splitViewRect
        }
    }

    func showTabbar() {
        if !tabbarVisibility {
            tabbarVisibility = true
            tabbar!.frame = tabbarRect
            splitView!.frame = splitViewRect
        }
    }
    
    func flip() {
        var doneRect = NSZeroRect
        doneRect.size.width = 105.0
        doneRect.size.height = 24.0
        let doneButton = SBBLKGUIButton(frame: doneRect)
        doneButton.title = NSLocalizedString("Done", comment: "")
        doneButton.target = self
        doneButton.action = "doneFlip"
        doneButton.enabled = true
        doneButton.keyEquivalent = "\r"
        flip(doneButton)
    }
    
    func flip(view: NSView) {
        var br = frame
        br.size.width = kSBBackWindowFrameWidth
        br.size.height = kSBBackWindowFrameHeight
        br.origin.x = frame.origin.x + (frame.size.width - br.size.width) / 2
        br.origin.y = frame.origin.y + (frame.size.height - br.size.height) / 2;
        br.size.height -= 23.0
        backWindow = NSWindow(contentRect: br, styleMask: (NSTitledWindowMask | NSClosableWindowMask), backing: .Buffered, defer: true)
        backWindow!.backgroundColor = SBWindowBackColor
        backWindow!.releasedWhenClosed = false
        view.frame = NSMakeRect((br.size.width - view.frame.size.width) / 2, (br.size.height - view.frame.size.height) / 2, view.frame.size.width, view.frame.size.height)
        backWindow!.contentView.addSubview(view)
        backWindow!.makeKeyAndOrderFront(nil)
        alphaValue = 0
    }
    
    func doneFlip() {
        backWindow?.close()
        backWindow = nil
        alphaValue = 1
    }
}